src/librustc_trans/collector.rs

```diff
@@ -304,6 +304,7 @@ fn collect_roots<'a, 'tcx>(scx: &SharedCrateContext<'a, 'tcx>,
         scx.tcx().hir.krate().visit_all_item_likes(&mut visitor);
     }
 
+    roots.retain(|root| root.is_instantiable(scx.tcx()));
```

Comment:
Hmm. This seems perfectly reasonable, but perhaps not optimal? (Also, why do we check below for whether something is instantiable, but not other places?)

Put another way, it might be better to have some helper routine for pushing into `self.output` that checks for "is instantiable" at that point -- perhaps this would also avoid enumerating the things reachable from a non-instantiable thing?

However, I see no obvious reason that this code is wrong.